From 12316ee8f4b3eb6a4195f78f71ac14b00162326b Mon Sep 17 00:00:00 2001
From: "Wu, Hao" <hao.wu@intel.com>
Date: Mon, 8 Dec 2014 18:08:47 +0800
Subject: [PATCH 013/441] phy-intel-cht: monitor USB-Host cable from extcon

This patch adds code to monitor on extcon dev for USB Host
cable status, then to switch to host/device mode by phy
mux.

Change-Id: Ie33b047e7a706c5ff8717323c0a62c204ab8ddb6
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-12562
Signed-off-by: Wu, Hao <hao.wu@intel.com>
---
 drivers/usb/phy/phy-intel-cht.c | 44 ++++++++++++++++++++++++++++++++++++++++-
 drivers/usb/phy/phy-intel-cht.h |  2 ++
 2 files changed, 45 insertions(+), 1 deletion(-)

--- a/drivers/usb/phy/phy-intel-cht.c
+++ b/drivers/usb/phy/phy-intel-cht.c
@@ -27,6 +27,7 @@
 #include <linux/usb.h>
 #include <linux/usb/gadget.h>
 #include <linux/usb/hcd.h>
+#include <linux/extcon.h>
 
 #include "../host/xhci.h"
 #include "../host/xhci-intel-cap.h"
@@ -42,6 +43,15 @@ static const char driver_name[] = "intel
 
 static struct cht_otg *cht_otg_dev;
 
+static int edev_state_to_id(struct extcon_dev *evdev)
+{
+	int state;
+
+	state = extcon_get_cable_state(evdev, "USB-Host");
+
+	return !evdev->state;
+}
+
 static int cht_otg_set_id_mux(struct cht_otg *otg_dev, int id)
 {
 	struct usb_bus *host = otg_dev->phy.otg->host;
@@ -186,7 +196,11 @@ static int cht_otg_start(struct platform
 	fsm->otg = otg_dev->phy.otg;
 	mutex_init(&fsm->lock);
 
-	fsm->id = 1;
+	/* Initialize the id value from extcon dev or default 1 */
+	if (otg_dev->cable_nb.edev)
+		fsm->id = edev_state_to_id(otg_dev->cable_nb.edev);
+	else
+		fsm->id = 1;
 	otg_statemachine(fsm);
 
 	dev_dbg(&pdev->dev, "initial ID pin set to %d\n", fsm->id);
@@ -386,6 +400,25 @@ static ssize_t store_otg_id(struct devic
 }
 static DEVICE_ATTR(otg_id, S_IWUSR|S_IWGRP, NULL, store_otg_id);
 
+static int cht_handle_extcon_otg_event(struct notifier_block *nb,
+					unsigned long event, void *param)
+{
+	struct extcon_dev *edev = param;
+	int id = edev_state_to_id(edev);
+
+	if (!cht_otg_dev)
+		return NOTIFY_DONE;
+
+	dev_info(cht_otg_dev->phy.dev, "[extcon notification]: USB-Host: %s\n",
+			id ? "Disconnected" : "connected");
+
+	/* update id value and schedule fsm work to start/stop host per id */
+	cht_otg_dev->fsm.id = id;
+	schedule_work(&cht_otg_dev->fsm_work);
+
+	return NOTIFY_OK;
+}
+
 static int cht_otg_probe(struct platform_device *pdev)
 {
 	struct cht_otg *cht_otg;
@@ -433,6 +466,13 @@ static int cht_otg_probe(struct platform
 	cht_otg_dev->nb.notifier_call = cht_otg_handle_notification;
 	usb_register_notifier(&cht_otg_dev->phy, &cht_otg_dev->nb);
 
+	/* Register on extcon for OTG event too */
+	cht_otg_dev->id_nb.notifier_call = cht_handle_extcon_otg_event;
+	status = extcon_register_interest(&cht_otg_dev->cable_nb, NULL,
+					"USB-Host", &cht_otg_dev->id_nb);
+	if (status)
+		dev_warn(&pdev->dev, "failed to register extcon notifier\n");
+
 	/* init otg-fsm */
 	status = cht_otg_start(pdev);
 	if (status) {
@@ -466,6 +506,7 @@ err3:
 	device_remove_file(&pdev->dev, &dev_attr_cht_otg_state);
 err2:
 	cht_otg_stop(pdev);
+	extcon_unregister_interest(&cht_otg_dev->cable_nb);
 	usb_remove_phy(&cht_otg_dev->phy);
 err1:
 	kfree(cht_otg->phy.otg);
@@ -480,6 +521,7 @@ static int cht_otg_remove(struct platfor
 	device_remove_file(&pdev->dev, &dev_attr_cht_otg_state);
 
 	cht_otg_stop(pdev);
+	extcon_unregister_interest(&cht_otg_dev->cable_nb);
 	usb_remove_phy(&cht_otg_dev->phy);
 
 	kfree(cht_otg_dev->phy.otg);
--- a/drivers/usb/phy/phy-intel-cht.h
+++ b/drivers/usb/phy/phy-intel-cht.h
@@ -25,6 +25,8 @@ struct cht_otg {
 	struct usb_phy phy;
 	struct otg_fsm fsm;
 	struct notifier_block nb;
+	struct extcon_specific_cable_nb cable_nb;
+	struct notifier_block id_nb;
 	struct work_struct fsm_work;
 	void __iomem *regs;
 };
